src/runtime/transport.ts

```diff
@@ -0,0 +1,190 @@
+import {
+  toConnectedEvent,
+  toMessageEvent,
+  createMessageEvent,
+  toRealTimeMessage,
+  createHeartbeatEvent,
+  createDisconnectEvent,
+  Encoding,
+} from "./messages";
+import { EventEmitter } from "node:events";
+import * as proto from "../proto/server/v1/realtime";
+import { ErrorEvent } from "ws";
+
+type MessageEventListener = (MessageEvent: proto.MessageEvent) => void;
+
+export type Newable<T> = { new (...args: any[]): T };
+
+interface TransportConfig {
+  heartbeatTimeout: number;
+  WebSocket: Newable<WebSocket>;
+}
+
+interface Session {
+  sessionId: string;
+  socketId: string;
+}
+
+type ConnectionState =
+  | "failed"
+  | "connecting"
+  | "connected"
+  | "uninitialized"
+  | "closing"
+  | "closed";
+
+export class Transport extends EventEmitter {
+  private channelListeners: Map<string, MessageEventListener[]>;
+  private ws: WebSocket;
+  private session?: Session;
+  private _isConnected: Promise<void>;
+  private _connectionState: ConnectionState = "uninitialized";
+  private heartbeatId: number | NodeJS.Timeout;
+  private config: TransportConfig;
+  private connectionResolved: (value: void | PromiseLike<void>) => void;
+  private encoding: Encoding;
+
+  constructor(config: TransportConfig) {
+    super();
+
+    this.config = config;
+    this.channelListeners = new Map();
+    this.heartbeatId = 0;
+    this.encoding = Encoding.json;
```

Comment:
this is good but just a note that default would be msgpack.

Reply:
Great. Once its implemented on the server I can switch to it on the client.

Reply:
I've switched to msgpack.